File: src-tauri/src/commands/gold_prices.rs
use super::{ApiResponse, DbPool};
use crate::models::{GoldPrice, SetGoldPriceRequest};
use tauri::State;

#[tauri::command]
pub async fn get_today_prices(
    pool: State<'_, DbPool>,
) -> Result<ApiResponse<Vec<GoldPrice>>, String> {
    let today = chrono::Local::now().format("%Y-%m-%d").to_string();

    let prices: Vec<GoldPrice> = sqlx::query_as::<_, GoldPrice>(
        r#"
        SELECT id, date, gold_type, purity, buy_price, sell_price, source, created_at
        FROM gold_prices
        WHERE date = ?
        ORDER BY gold_type, purity DESC
        "#,
    )
    .bind(&today)
    .fetch_all(&pool.0)
    .await
    .map_err(|e| e.to_string())?;

    Ok(ApiResponse::success(prices))
}

#[tauri::command]
pub async fn set_gold_price(
    pool: State<'_, DbPool>,
    request: SetGoldPriceRequest,
) -> Result<ApiResponse<GoldPrice>, String> {
    let today = chrono::Local::now().format("%Y-%m-%d").to_string();
    let id = uuid::Uuid::new_v4().to_string();

    // Upsert price
    sqlx::query(
        r#"
        INSERT INTO gold_prices (id, date, gold_type, purity, buy_price, sell_price, source)
        VALUES (?, ?, ?, ?, ?, ?, 'Manual')
        ON CONFLICT(date, gold_type, purity)
        DO UPDATE SET buy_price = excluded.buy_price, sell_price = excluded.sell_price
        "#,
    )
    .bind(&id)
    .bind(&today)
    .bind(&request.gold_type)
    .bind(request.purity)
    .bind(request.buy_price)
    .bind(request.sell_price)
    .execute(&pool.0)
    .await
    .map_err(|e| e.to_string())?;

    let price: GoldPrice = sqlx::query_as::<_, GoldPrice>(
        r#"
        SELECT id, date, gold_type, purity, buy_price, sell_price, source, created_at
        FROM gold_prices
        WHERE date = ? AND gold_type = ? AND purity = ?
        "#,
    )
    .bind(&today)
    .bind(&request.gold_type)
    .bind(request.purity)
    .fetch_one(&pool.0)
    .await
    .map_err(|e| e.to_string())?;

    Ok(ApiResponse::success(price))
}

#[tauri::command]
pub async fn get_price_history(
    pool: State<'_, DbPool>,
    gold_type: String,
    purity: i32,
    days: i32,
) -> Result<ApiResponse<Vec<GoldPrice>>, String> {
    let prices: Vec<GoldPrice> = sqlx::query_as::<_, GoldPrice>(
        r#"
        SELECT id, date, gold_type, purity, buy_price, sell_price, source, created_at
        FROM gold_prices
        WHERE gold_type = ? AND purity = ?
        ORDER BY date DESC
        LIMIT ?
        "#,
    )
    .bind(&gold_type)
    .bind(purity)
    .bind(days)
    .fetch_all(&pool.0)
    .await
    .map_err(|e| e.to_string())?;

    Ok(ApiResponse::success(prices))
}

#[tauri::command]
pub async fn get_all_prices_for_date(
    pool: State<'_, DbPool>,
    date: String,
) -> Result<ApiResponse<Vec<GoldPrice>>, String> {
    let prices: Vec<GoldPrice> = sqlx::query_as::<_, GoldPrice>(
        r#"
        SELECT id, date, gold_type, purity, buy_price, sell_price, source, created_at
        FROM gold_prices
        WHERE date = ?
        ORDER BY gold_type, purity DESC
        "#,
    )
    .bind(&date)
    .fetch_all(&pool.0)
    .await
    .map_err(|e| e.to_string())?;

    Ok(ApiResponse::success(prices))
}

#[tauri::command]
pub async fn get_price_for_calculation(
    pool: State<'_, DbPool>,
    gold_type: String,
    purity: i32,
) -> Result<ApiResponse<Option<GoldPrice>>, String> {
    let today = chrono::Local::now().format("%Y-%m-%d").to_string();

    let price: Option<GoldPrice> = sqlx::query_as::<_, GoldPrice>(
        r#"
        SELECT id, date, gold_type, purity, buy_price, sell_price, source, created_at
        FROM gold_prices
        WHERE date = ? AND gold_type = ? AND purity = ?
        "#,
    )
    .bind(&today)
    .bind(&gold_type)
    .bind(purity)
    .fetch_optional(&pool.0)
    .await
    .map_err(|e| e.to_string())?;

    Ok(ApiResponse::success(price))
}

File: src/api/goldPrices.ts
/**
 * Gold Prices API - Gold price management commands
 */

import { tauriInvoke, ApiResponse } from './index';
import type { GoldPrice } from '../types';

export interface SetGoldPriceRequest {
  gold_type: 'LM' | 'UBS' | 'Lokal';
  purity: number;
  buy_price: number;
  sell_price: number;
}

/**
 * Get today's gold prices
 */
export async function getTodayPrices(): Promise<ApiResponse<GoldPrice[]>> {
  return tauriInvoke<GoldPrice[]>('get_today_prices');
}

/**
 * Set or update gold price for today
 */
export async function setGoldPrice(request: SetGoldPriceRequest): Promise<ApiResponse<GoldPrice>> {
  return tauriInvoke<GoldPrice>('set_gold_price', { request });
}

/**
 * Get price history for a specific gold type and purity
 */
export async function getPriceHistory(
  goldType: string,
  purity: number,
  days: number = 30
): Promise<ApiResponse<GoldPrice[]>> {
  return tauriInvoke<GoldPrice[]>('get_price_history', {
    goldType,
    purity,
    days,
  });
}

/**
 * Get all gold prices for a specific date
 */
export async function getAllPricesForDate(date: string): Promise<ApiResponse<GoldPrice[]>> {
  return tauriInvoke<GoldPrice[]>('get_all_prices_for_date', { date });
}

/**
 * Get today's price for a specific gold type and purity (for calculations)
 */
export async function getPriceForCalculation(
  goldType: string,
  purity: number
): Promise<ApiResponse<GoldPrice | null>> {
  return tauriInvoke<GoldPrice | null>('get_price_for_calculation', {
    goldType,
    purity,
  });
}
